arangod/Replication2/ReplicatedLog/LogLeader.cpp

```diff
@@ -0,0 +1,945 @@
+
+    LOG_CTX("8696f", INFO, _logContext) << "resign";
+    leaderData._didResign = true;
+    static_assert(
+        std::is_nothrow_constructible_v<DeferredAction, std::add_rvalue_reference_t<decltype(action)>>);
+    static_assert(noexcept(std::declval<LocalFollower&&>().resign()));
+    return std::make_tuple(std::move(localFollower).resign(),
+                           DeferredAction(std::move(action)));
+  });
+}
+
+auto replicated_log::LogLeader::readReplicatedEntryByIndex(LogIndex idx) const
+    -> std::optional<PersistingLogEntry> {
+  return _guardedLeaderData.doUnderLock([&idx](auto& leaderData) -> std::optional<PersistingLogEntry> {
+    if (leaderData._didResign) {
+      THROW_ARANGO_EXCEPTION(TRI_ERROR_REPLICATION_REPLICATED_LOG_LEADER_RESIGNED);
+    }
+    if (auto entry = leaderData._inMemoryLog.getEntryByIndex(idx);
+        entry.has_value() && entry->entry().logIndex() <= leaderData._commitIndex) {
+      return entry->entry();
+    } else {
+      return std::nullopt;
+    }
+  });
+}
+
+auto replicated_log::LogLeader::getStatus() const -> LogStatus {
+  return _guardedLeaderData.doUnderLock([term = _currentTerm](auto& leaderData) {
+    if (leaderData._didResign) {
+      THROW_ARANGO_EXCEPTION(TRI_ERROR_REPLICATION_REPLICATED_LOG_LEADER_RESIGNED);
+    }
+    LeaderStatus status;
+    status.local = leaderData.getLocalStatistics();
+    status.term = term;
+    for (FollowerInfo const& f : leaderData._follower) {
+      status.follower[f._impl->getParticipantId()] = {
+          LogStatistics{f.lastAckedEntry, f.lastAckedCommitIndex}, f.lastErrorReason,
+          std::chrono::duration_cast<std::chrono::duration<double, std::milli>>(f._lastRequestLatency)
+              .count()};
+    }
+    return LogStatus{std::move(status)};
+  });
+}
+
+auto replicated_log::LogLeader::insert(LogPayload payload, bool waitForSync) -> LogIndex {
+  auto index = insert(std::move(payload), waitForSync, doNotTriggerAsyncReplication);
+  triggerAsyncReplication();
+  return index;
+}
+
+auto replicated_log::LogLeader::insert(LogPayload payload, [[maybe_unused]] bool waitForSync,
+                                       DoNotTriggerAsyncReplication) -> LogIndex {
+  // TODO Handle waitForSync!
+  auto const insertTp = InMemoryLogEntry::clock::now();
+  // Currently we use a mutex. Is this the only valid semantic?
+  return _guardedLeaderData.doUnderLock([&](GuardedLeaderData& leaderData) {
+    if (leaderData._didResign) {
+      THROW_ARANGO_EXCEPTION(TRI_ERROR_REPLICATION_REPLICATED_LOG_LEADER_RESIGNED);
+    }
+    auto const index = leaderData._inMemoryLog.getNextIndex();
+    auto const payloadSize = payload.byteSize();
+    auto logEntry =
+        InMemoryLogEntry(PersistingLogEntry(_currentTerm, index, std::move(payload)));
+    logEntry.setInsertTp(insertTp);
+    leaderData._inMemoryLog.appendInPlace(_logContext, std::move(logEntry));
+    _logMetrics->replicatedLogInsertsBytes->count(payloadSize);
+    return index;
+  });
```

Comment:
I guess this section is critical for the log write throughput.
We could do various things to speed it up:
* calculate the `payloadSize` outside of the mutex
* update the metrics for inserted bytes outside of the mutex
* create an empty log entry with the payload and `insertTp` outside of the mutex, and then later set these attributes under the mutex.

Reply:
Good idea, I will change that. Though iirc we never had contention here in our performance tests so far. We planned to evaluate if we can change this to a lock-free implementation if this turns out to be a bottleneck, but it didn't seem necessary so far.